add serverless.ts support

bryan-hunter opened the pull request:
<!-- ⚠️⚠️ Acknowledge ALL below remarks -->
<!-- ⚠️⚠️ PR will not be processed if it doesn't meet outlined criteria -->

<!-- ⚠️⚠️ Do not propose PR's without prior agreement on solution in corresponding issue -->
<!-- ⚠️⚠️ Only documentation updates and obvious bug fixes are welcome without it -->

<!--
⚠️⚠️ Ensure to follow code style guidelines
https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md#code-style
-->

<!--
⚠️⚠️ Ensure to cover changes with tests written according to test guidelines
https://github.com/serverless/serverless/blob/master/tests/README.md
-->

<!-- ⚠️⚠️ Ensure that support for Node.js v6 is maintained. -->

<!--
⚠️⚠️ Ensure that proposed change passes CI. Confirm on that by running following scripts:
• npm run prettier-check
• npm run lint
• npm test
-->

<!--
⚠️⚠️ If proposed change touches integration with AWS services, confirm integration tests pass:
https://github.com/serverless/serverless/blob/master/tests/README.md#aws-integration-tests
-->

<!-- ⚠️⚠️ After your PR is submitted, review the final CI status and address eventual failure -->

<!-- ⚠️⚠️ Answer below questions -->

<!--
Q1: Provide link to corresponding issue

Note: Remove this section if it's documentation update or obvious bug fix that has no corresponding issue. In such case provide a short description of made changes
-->

Closes: https://github.com/serverless/serverless/issues/6335

Very simple PR to handle `serverless.ts` files.

Try to find `ts-node` module when `.ts` file is detected and throw an error that you need to provide this dependency to use serverless.ts files.

bryan-hunter commented:
yesterday I added AWS types for `serverless.yml` files https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44894

the combination of these 2 things could make it very easy to use Serverless with AWS and get type safety/auto completion.

medikoo commented on lib/utils/getServerlessConfigFile.js:
```diff
@@ -46,28 +50,34 @@ const getServerlessConfigFilePath = serverless => {
-    // use require to load serverless.js
+    if (isTs) {
+      // eslint-disable-next-line global-require
+      require('ts-node').register();
```
I would prefer not add it as dependency. I think TS users should just ensure that both `typescript` and `ts-node` are installed in their setup.

Also we need to ensure it'll work with _standalone_ build, which makes it a bit more tricky (as it doesn't have natural access to _service_ or _global_ npm dependencies)

I imagine reliable logic of acquisition of `ts-node` as follows:

```javascript
const resolveModulePath = require('ncjsm/resolve');
const spawn = require('child-process-ext/spawn')
...

const tsNode = () => {
	// 1. Resolve as serverless peer dependency
	return resolveModulePath(__dirname, "ts-node")
		.catch((error) => {
			if (error.code !== "MODULE_NOT_FOUND") throw error;
			// 2. Resolve as service dependency
			return resolveModulePath(path.dirname(configFile), "ts-node");
		})
		.catch((error) => {
			if (error.code !== "MODULE_NOT_FOUND") throw error;
			// 2. Resolve as global installation
			return spawn("npm", ["root", "-g"]).then(
				({ stdoutBuffer }) => {
					return resolveModulePath("/", `${String(stdoutBuffer).trim()}/ts-node`).catch(
						(error) => {
							if (error.code !== "MODULE_NOT_FOUND") throw error;
							return null;
						}
					);
				},
				(error) => null
			);
		})
		.then((tsNodePath) => {
			if (!tsNodePath) {
				throw new ServerlessError(
					"Ensure 'ts-node' dependency for working with TypeScript configuration files"
				);
			}
			return require(tsNodePath);
		});
};
```

bryan-hunter replied:
I updated this to remove ts-node and resolve it here - some weird stuff I had to do to get it to resolve symlinks (because it returns an object instead of filepath), but works for all 3 cases.  Let me know if you like this approach, or if you want to include ts-node as a dep.

bryan-hunter commented on lib/utils/getServerlessConfigFile.js:
```diff
@@ -46,28 +50,34 @@ const getServerlessConfigFilePath = serverless => {
+const handleJsOrTsConfigFile = (configFile, isTs) =>
   BbPromise.try(() => {
-    // use require to load serverless.js
+    if (isTs) {
```
updated

bryan-hunter commented on lib/utils/getServerlessConfigFile.js:
```diff
@@ -46,28 +50,34 @@ const getServerlessConfigFilePath = serverless => {
       return config;
     }
-    throw new Error('serverless.js must export plain object');
+    throw new Error(`serverless.${isTs ? 'ts' : 'js'} must export plain object`);
```
updated

bryan-hunter commented on lib/utils/getServerlessConfigFile.js:
```diff
@@ -46,28 +50,34 @@ const getServerlessConfigFilePath = serverless => {
   BbPromise.try(() => {
-    // use require to load serverless.js
+    if (isTs) {
+      // eslint-disable-next-line global-require
```
removed

bryan-hunter commented on package.json:
```diff
@@ -171,6 +171,7 @@
     "sinon-chai": "^3.5.0",
     "standard-version": "^8.0.0",
     "strip-ansi": "^5.2.0",
+    "ts-node": "^8.10.1",
```
added as devDependency for test case

medikoo replied:
Let's not add it.

In previous comment I proposed to use fixture and when testing in fixture we may create a dummy `node_modules/ts-node.js` in a fixture folder, which should be a module that exposes a `register` method. Through that we may confirm it doesn't crash and that `register` is invoked

bryan-hunter commented:
@medikoo I have removed `ts-node` as a dependency (made it a devDependency for testing).

I have updated code and added tests.  Let me know what you think :)

medikoo commented on lib/utils/getServerlessConfigFile.test.js:
```diff
@@ -146,6 +148,81 @@ describe('#getServerlessConfigFile()', () => {
     ).to.be.rejectedWith('serverless.js must export plain object');
   });
 
+  it('should return the file content if a serverless.ts file found', () => {
```
For tests we should use `runServerless` util, as documented here: https://github.com/serverless/serverless/blob/master/tests/README.md#unit-tests

It's the only reliable way to confirm that it works. We may prepare some basic fixture that involves `serverless.ts` config (but with no real TS internals, to not impose TS dependency on test engine)

bryan-hunter replied:
I added a fixture and added a test case.  Let me know if you think there's more that I should be testing

bryan-hunter commented on lib/utils/getServerlessConfigFile.js:
```diff
@@ -46,28 +53,68 @@ const getServerlessConfigFilePath = serverless => {
+        throw error;
+      }
+      // if there is a symlink, we need to get 'realPath' from the return object
+      const pathString = typeof result === 'string' ? result : result.realPath;
```
updated

bryan-hunter commented on lib/utils/getServerlessConfigFile.js:
```diff
@@ -46,28 +53,68 @@ const getServerlessConfigFilePath = serverless => {
+        const error = new Error(`Cannot find module '${args[1]}'`);
+        error.code = 'MODULE_NOT_FOUND';
+        throw error;
+      }
```
removed

bryan-hunter commented on lib/utils/getServerlessConfigFile.js:
```diff
@@ -46,28 +53,68 @@ const getServerlessConfigFilePath = serverless => {
+      }
+      // if there is a symlink, we need to get 'realPath' from the return object
+      const pathString = typeof result === 'string' ? result : result.realPath;
+      return require(pathString);
```
moved it to after the chain of `catch`es once module is resolved, so won't error on require and be caught erroneously

bryan-hunter commented on lib/utils/getServerlessConfigFile.js:
```diff
@@ -46,28 +53,68 @@ const getServerlessConfigFilePath = serverless => {
+  const resolveAsServiceDependency = () => requireOrThrow(serviceDir, 'ts-node');
+  const resolveAsGlobalInstallation = () =>
+    spawn('npm', ['root', '-g']).then(({ stdoutBuffer }) => {
+      return requireOrThrow('/', `${String(stdoutBuffer).trim()}/ts-node`);
```
modified to use `require.resolve`

bryan-hunter commented on lib/utils/getServerlessConfigFile.js:
```diff
@@ -46,28 +53,68 @@ const getServerlessConfigFilePath = serverless => {
     getServerlessConfigFilePath(serverless).then(configFilePath => {
       if (configFilePath !== null) {
-        const isJSConfigFile = _.last(_.split(configFilePath, '.')) === 'js';
+        const fileExtension = _.last(_.split(configFilePath, '.'));
```
modified to use `path.extname` and changed the if check to include the `.` on `.ts` and `.js`

medikoo commented on lib/utils/getServerlessConfigFile.js:
```diff
@@ -46,28 +53,58 @@ const getServerlessConfigFilePath = serverless => {
+  const resolveModuleRealPath = (...args) =>
+    resolveModulePath(...args).then(({ realPath }) => realPath);
+
+  const createOnError = cb => error => {
```
Let's rename it to `ifNotFoundContinueWith` (at least `createOnError` doesn't reflect what this util does)

bryan-hunter replied:
done

medikoo commented on lib/utils/getServerlessConfigFile.js:
```diff
@@ -46,28 +53,58 @@ const getServerlessConfigFilePath = serverless => {
+  const throwTsNodeError = () => {
+    throw new ServerlessError(
+      'Ensure "ts-node" dependency when working with TypeScript configuration files'
+    );
```
For better processing (e.g. in tests), it'll be nice to add _code_ to this error (it's accepted as second argument), e.g. `'TS_NODE_NOT_FOUND'`

bryan-hunter replied:
done and used in test :)

medikoo commented on tests/fixtures/serverlessTs/node_modules/ts-node.js:
```diff
@@ -0,0 +1,3 @@
+'use strict';
+
+module.exports.register = () => null;
```
I was thinking of adding this conditionally.

So first test without it (and confirm that `'TS_NODE_NOT_FOUND'` error was thrown), then create this file in test, and confirm it resolves.

Note that `fixures.cleanup` will remove created file from fixture, so you don't need to worry about cleanup

bryan-hunter replied:
done.  I did have to pass in "extraPaths" option to fixture.cleanup because it only tries to cleanup `.serverless`, but was easy enough and seems to work nicely.

Also had to pass in `shouldStubSpawn: true` so wouldn't resolve my global npm ts-node... was nice to already have that option available - nice job on making everything easy to use though :)

bryan-hunter commented:
@medikoo do you know when it can be released?

medikoo commented:
@bryan-hunter tomorrow morning CET time

DavidEspinosa42 commented:
This would allow to replace a current serverless.yml for a serverless.ts?
Can you provide a simple example of the .ts file format?

bryan-hunter commented:
> This would allow to replace a current serverless.yml for a serverless.ts?
> Can you provide a simple example of the .ts file format?

@DavidEspinosa42 this would be one example:

`serverless.ts`
```
import type Aws from 'serverless/aws';

const myService: Aws.Serverless = {
    plugins: ['serverless-webpack', 'serverless-offline'],
    frameworkVersion: '>=1.72.0',
    service: {
        name: 'YOUR_SERVICE',
    },
    provider: {
        name: 'aws',
        runtime: 'nodejs10.x',
        timeout: 30,
        memorySize: 1024,
        environment: {
            SERVICE_NAME: 'YOUR_SERVICE',
        },
        apiGateway: {
            minimumCompressionSize: 0,
        },
        tracing: {
            lambda: true,
            apiGateway: true,
        },
        logs: {
            frameworkLambda: true,
        },
    },
    package: {
        individually: true,
    },
    custom: {
        webpack: {
            includeModules: {
                forceExclude: ['aws-sdk'],
            },
            packager: 'yarn',
        },
        'serverless-offline': {
            host: '0.0.0.0',
            port: 3000,
            location: '.webpack/service',
            apiKey: 'fakeKey',
        },
    },
    resources: {
        Resources: {
            GatewayResponseDefault4XX: {
                Type: 'AWS::ApiGateway::GatewayResponse',
                Properties: {
                    ResponseParameters: {
                        'gatewayresponse.header.Access-Control-Allow-Origin': "'*'",
                        'gatewayresponse.header.Access-Control-Allow-Headers': "'*'",
                    },
                    ResponseType: 'DEFAULT_4XX',
                    RestApiId: {
                        Ref: 'ApiGatewayRestApi',
                    },
                    ResponseTemplates: {
                        'application/json':
                            '{"error": {"code": "custom-4XX-generic", "message": $context.error.messageString, "type": "$context.error.responseType"}, "requestId": "$context.requestId"}',
                    },
                },
            },
            GatewayResponseDefault5XX: {
                Type: 'AWS::ApiGateway::GatewayResponse',
                Properties: {
                    ResponseParameters: {
                        'gatewayresponse.header.Access-Control-Allow-Origin': "'*'",
                        'gatewayresponse.header.Access-Control-Allow-Headers': "'*'",
                    },
                    ResponseType: 'DEFAULT_5XX',
                    RestApiId: {
                        Ref: 'ApiGatewayRestApi',
                    },
                    ResponseTemplates: {
                        'application/json':
                            '{"error": {"code": "custom-5XX-generic", "message": $context.error.messageString, "type": "$context.error.responseType"}, "requestId": "$context.requestId"}',
                    },
                },
            },
            DynamoDbTable: {
                Type: 'AWS::DynamoDB::Table',
                DeletionPolicy: 'Retain',
                Properties: {
                    TableName: 'YOUR_TABLE_NAME',
                    BillingMode: 'PAY_PER_REQUEST',
                    AttributeDefinitions: [
                        {
                            AttributeName: 'id',
                            AttributeType: 'S',
                        },
                    ],
                    KeySchema: [
                        {
                            AttributeName: 'id',
                            KeyType: 'HASH',
                        },
                    ],
                    TimeToLiveSpecification: {
                        AttributeName: 'expiresAt',
                        Enabled: true,
                    },
                },
            },
        },
    },
    functions: {
        get: {
            handler: 'src/handler/YourHandler.get',
            environment: {},
            events: [
                {
                    http: {
                        path: 'data/{id}',
                        method: 'get',
                        private: true
                    },
                },
            ],
            tags: {},
        },
    },
};

module.exports = myService;
```
be sure to install @types/serverless (some of the types may be slightly incorrect there, but can easily make a PR to fix them) if you want to use Serverless types out of the box - otherwise you can use your own typings :)

DavidEspinosa42 commented:
Thanks Bryan!! 
It worked perfectly 😄 
Awesome work!

bharat-tiwari commented:
how to use the custom variables in the ts version

like, how would the below `yml` would convert to `ts`:


```
…
…
 custom:
    dbTable: 'myDBTable-${opt:stage}'
…
…
  environment:
    DB_TABLE: ${self:custom.dbTable}

```

DavidEspinosa42 commented:
@bharat-tiwari You could use something like yargs:

import * as yargs from 'yargs';

custom: {
	'dbTable': 'myDBTable-' + yargs.argv.stage
}

mospina commented:
I still can figure it out how to do this:

```
environment:
    DB_TABLE: ${self:custom.dbTable}
```

DavidEspinosa42 commented:
@mospina have you tried with quotes?
DB_TABLE: '${self:custom.dbTable}'
